[aotd] No AOT compilation for backward

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136675

For pre-dispatch tracing we must not do AOT compilation of BW, before we see the real tangents.

IvanKobzarev committed:
[wip] No AOT compilation for backward

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[wip] No AOT compilation for backward"




[ghstack-poisoned]

bdhirsh commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1951,13 +1956,30 @@ def call_compiled_backward():
+                                )
+                                log.warning(
+                                    "failed to eagerly compile backwards for dynamic, suppressing in case backwards not needed",
+                                    exc_info=True,
```
another good E2E test would be something like this, where there are actually extra guards that inductor adds during backward compilation:
```
import torch

class MyFunc(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x, y):
        return x, y

    @staticmethod
    def backward(ctx, grad_x, grad_y):
        return grad_x + grad_x, grad_y + grad_y
        tmp = torch.cat([grad_x, grad_y])
        out = tmp + tmp
        return grad_x, out.sum().expand(saved_y.shape)

def f(x, y):
    return MyFunc.apply(x, y)


f_compiled = torch.compile(f, dynamic=True)

x_ref = torch.randn(6, device='cuda', dtype=torch.float16, requires_grad=True)
y_ref = torch.randn(4, 5, device='cuda', dtype=torch.float16, requires_grad=True)
x_test = x_ref.clone().detach().requires_grad_()
y_test = y_ref.clone().detach().requires_grad_()

out_ref = f(x_ref, y_ref)
(out_ref[0].sum() + out_ref[1].sum()).sum().backward()

out_test = f_compiled(x_test, y_test)
(out_test[0].sum() + out_test[1].sum()).sum().backward()

x_ref2 = torch.randn(6, device='cuda', dtype=torch.float16, requires_grad=True)
y_ref2 = torch.randn(47000, 47001, device='cuda', dtype=torch.float16, requires_grad=True)
x_test2 = x_ref2.clone().detach().requires_grad_()
y_test2 = y_ref2.clone().detach().requires_grad_()

out_ref2 = f(x_ref2, y_ref2)
(out_ref2[0].sum() + out_ref2[1].sum()).sum().backward()

out_test2 = f_compiled(x_test2, y_test2)
(out_test2[0].sum() + out_test2[1].sum()).sum().backward()
# assert outputs and gradients are correct
```

IvanKobzarev committed:
Update on "[wip] No AOT compilation for backward"


For pre-dispatch tracing we must not do AOT compilation of BW, before we see the real tangents.




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] No AOT compilation for backward"


For pre-dispatch tracing we must not do AOT compilation of BW, before we see the real tangents.




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] No AOT compilation for backward"


For pre-dispatch tracing we must not do AOT compilation of BW, before we see the real tangents.




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] No AOT compilation for backward"


For pre-dispatch tracing we must not do AOT compilation of BW, before we see the real tangents.




[ghstack-poisoned]
